src/main/java/com/ibm/disni/util/MemoryAllocation.java

```diff
@@ -21,67 +21,142 @@
 
 package com.ibm.disni.util;
 
+import org.slf4j.Logger;
+
 import java.nio.ByteBuffer;
 import java.nio.ByteOrder;
+import java.util.Comparator;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.atomic.AtomicLong;
 
-public class MemoryAllocation {
-	public enum MemType {
-		INDIRECT, DIRECT,
-	}
 
+public class MemoryAllocation {
+	private static final int MIN_BLOCK_SIZE = 64; // 64B
+	private int MAX_CACHE_SIZE = 4 * 1024 * 1024; // 4MB
+	private final ConcurrentHashMap<Integer, AllocatorStack> allocStackMap =
+		new ConcurrentHashMap<>();
+	private static final Logger logger = DiSNILogger.getLogger();
 	private static MemoryAllocation instance = null;
-	private ConcurrentHashMap<String, MemBuf> table;
 
 	public synchronized static MemoryAllocation getInstance() {
 		if (instance == null) {
 			instance = new MemoryAllocation();
 		}
 		return instance;
 	}
-	
-	private MemoryAllocation() {
-		table = new ConcurrentHashMap<String, MemBuf>();
-	}
-	
-	public MemBuf allocate(int size, MemType type, String classname) {
-		if (type == MemType.DIRECT) {
-			String key = generateKey(size, type, classname);
-			MemBuf buf = table.remove(key);
-			if (buf != null) {
-				buf.getBuffer().clear();
-				return buf;
+
+	private class AllocatorStack {
```

Comment:
Please add a comment to explain what this is and how it is used - just a 2-3 lines summary